File: internal/browsernetworktopology/candidate.go
package browsernetworktopology

import (
	"errors"
	"fmt"
	"net/netip"
	"regexp"
	"strings"
)

var ErrInvalidCandidatePath = errors.New("invalid selected candidate path")

var candidateMDNSLabelPattern = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$`)

type SelectedPairObservation string

const (
	SelectedPairPresent SelectedPairObservation = "present"
	SelectedPairAbsent  SelectedPairObservation = "absent"
)

type CandidatePath struct {
	SelectedPair        SelectedPairObservation `json:"selectedPair"`
	LocalCandidateType  *CandidateType          `json:"localCandidateType"`
	LocalAddress        *string                 `json:"localAddress"`
	LocalPort           *uint16                 `json:"localPort"`
	RemoteCandidateType *CandidateType          `json:"remoteCandidateType"`
	RemoteAddress       *string                 `json:"remoteAddress"`
	RemotePort          *uint16                 `json:"remotePort"`
	Protocol            *TransportProtocol      `json:"protocol"`
}

type CandidatePolicyOutcome string

const (
	CandidatePolicyMatched      CandidatePolicyOutcome = "matched"
	CandidatePolicyMismatched   CandidatePolicyOutcome = "mismatched"
	CandidatePolicyNotEvaluated CandidatePolicyOutcome = "not-evaluated"
)

type CandidateRationaleCode string

const (
	RationaleSelectedPairRequired               CandidateRationaleCode = "selected-pair-required"
	RationaleSelectedPairProhibited             CandidateRationaleCode = "selected-pair-prohibited"
	RationaleLocalCandidateTypeForbidden        CandidateRationaleCode = "local-candidate-type-forbidden"
	RationaleLocalCandidateTypeNotAllowed       CandidateRationaleCode = "local-candidate-type-not-allowed"
	RationaleLocalCandidateTypeRequiredMissing  CandidateRationaleCode = "local-candidate-type-required-missing"
	RationaleRemoteCandidateTypeForbidden       CandidateRationaleCode = "remote-candidate-type-forbidden"
	RationaleRemoteCandidateTypeNotAllowed      CandidateRationaleCode = "remote-candidate-type-not-allowed"
	RationaleRemoteCandidateTypeRequiredMissing CandidateRationaleCode = "remote-candidate-type-required-missing"
	RationaleProtocolForbidden                  CandidateRationaleCode = "protocol-forbidden"
	RationaleProtocolNotAllowed                 CandidateRationaleCode = "protocol-not-allowed"
	RationaleProtocolRequiredMissing            CandidateRationaleCode = "protocol-required-missing"
)

func (path CandidatePath) Validate() error {
	switch path.SelectedPair {
	case SelectedPairAbsent:
		if path.LocalCandidateType != nil || path.LocalAddress != nil || path.LocalPort != nil ||
			path.RemoteCandidateType != nil || path.RemoteAddress != nil || path.RemotePort != nil ||
			path.Protocol != nil {
			return fmt.Errorf("%w: absent selected pair carries candidate fields", ErrInvalidCandidatePath)
		}
	case SelectedPairPresent:
		if path.LocalCandidateType == nil || path.RemoteCandidateType == nil || path.Protocol == nil ||
			path.RemoteAddress == nil || path.RemotePort == nil || *path.RemotePort == 0 ||
			!validCandidateType(*path.LocalCandidateType) ||
			!validCandidateType(*path.RemoteCandidateType) || !validProtocol(*path.Protocol) ||
			!validCandidateLocalAddress(path.LocalAddress) ||
			(path.LocalPort != nil && *path.LocalPort == 0) || !validCandidateIPAddress(*path.RemoteAddress) {
			return fmt.Errorf("%w: present selected pair lacks a valid candidate type or protocol", ErrInvalidCandidatePath)
		}
	default:
		return fmt.Errorf("%w: selected-pair observation is unknown", ErrInvalidCandidatePath)
	}
	return nil
}

func validCandidateLocalAddress(address *string) bool {
	if address == nil {
		return true
	}
	if validCandidateIPAddress(*address) {
		return true
	}
	if len(*address) > 253 || !strings.HasSuffix(*address, ".local") {
		return false
	}
	for label := range strings.SplitSeq(*address, ".") {
		if len(label) < 1 || len(label) > 63 || !candidateMDNSLabelPattern.MatchString(label) {
			return false
		}
	}
	return true
}

func validCandidateIPAddress(address string) bool {
	_, err := netip.ParseAddr(address)
	return err == nil
}

// EvaluateCandidatePath is browser-neutral: the same observed pair is classified
// identically regardless of which engine produced getStats evidence.
func EvaluateCandidatePath(
	policy CandidatePolicy,
	expectation ConnectivityExpectation,
	path CandidatePath,
) (CandidatePolicyOutcome, []CandidateRationaleCode, error) {
	if err := policy.Validate(expectation); err != nil {
		return "", nil, err
	}
	if err := path.Validate(); err != nil {
		return "", nil, err
	}

	if path.SelectedPair == SelectedPairAbsent {
		if policy.SelectedPair == SelectedPairRequired {
			return CandidatePolicyMismatched, []CandidateRationaleCode{RationaleSelectedPairRequired}, nil
		}
		return CandidatePolicyMatched, []CandidateRationaleCode{}, nil
	}
	if policy.SelectedPair == SelectedPairProhibited {
		return CandidatePolicyMismatched, []CandidateRationaleCode{RationaleSelectedPairProhibited}, nil
	}

	rationales := make([]CandidateRationaleCode, 0, 6)
	rationales = appendCandidateRationales(
		rationales,
		*path.LocalCandidateType,
		policy.LocalCandidateTypes,
		RationaleLocalCandidateTypeForbidden,
		RationaleLocalCandidateTypeNotAllowed,
		RationaleLocalCandidateTypeRequiredMissing,
	)
	rationales = appendCandidateRationales(
		rationales,
		*path.RemoteCandidateType,
		policy.RemoteCandidateTypes,
		RationaleRemoteCandidateTypeForbidden,
		RationaleRemoteCandidateTypeNotAllowed,
		RationaleRemoteCandidateTypeRequiredMissing,
	)
	rationales = appendProtocolRationales(rationales, *path.Protocol, policy.Protocols)
	if len(rationales) == 0 {
		return CandidatePolicyMatched, rationales, nil
	}
	return CandidatePolicyMismatched, rationales, nil
}

func appendCandidateRationales(
	rationales []CandidateRationaleCode,
	observed CandidateType,
	constraint CandidateTypeConstraint,
	forbiddenCode CandidateRationaleCode,
	notAllowedCode CandidateRationaleCode,
	requiredMissingCode CandidateRationaleCode,
) []CandidateRationaleCode {
	if containsCandidateType(constraint.Forbidden, observed) {
		rationales = append(rationales, forbiddenCode)
	} else if !containsCandidateType(constraint.Allowed, observed) {
		rationales = append(rationales, notAllowedCode)
	}
	if len(constraint.Required) == 1 && constraint.Required[0] != observed {
		rationales = append(rationales, requiredMissingCode)
	}
	return rationales
}

func appendProtocolRationales(
	rationales []CandidateRationaleCode,
	observed TransportProtocol,
	constraint ProtocolConstraint,
) []CandidateRationaleCode {
	if containsProtocol(constraint.Forbidden, observed) {
		rationales = append(rationales, RationaleProtocolForbidden)
	} else if !containsProtocol(constraint.Allowed, observed) {
		rationales = append(rationales, RationaleProtocolNotAllowed)
	}
	if len(constraint.Required) == 1 && constraint.Required[0] != observed {
		rationales = append(rationales, RationaleProtocolRequiredMissing)
	}
	return rationales
}

func exactRationaleCodes(actual, expected []CandidateRationaleCode) bool {
	if len(actual) != len(expected) {
		return false
	}
	for index := range expected {
		if actual[index] != expected[index] {
			return false
		}
	}
	return true
}

File: web/scripts/browser-network-matrix/candidate.ts
import type { NetworkMatrixVocabularyConstraint, NetworkTopologyProfile } from './profile.ts'
import { isIP } from 'node:net'
import {
  NETWORK_MATRIX_CANDIDATE_TYPES,
  NETWORK_MATRIX_PROTOCOLS,
  NETWORK_MATRIX_SELECTED_PAIR_OBSERVATIONS,
  type NetworkMatrixCandidatePolicyOutcome,
  type NetworkMatrixCandidateType,
  type NetworkMatrixProtocol,
  type NetworkMatrixRationaleCode,
} from './vocabulary.ts'
import {
  networkMatrixError,
  requireEnum,
  requireExactKeys,
  requireRecord,
} from './contract-support.ts'

export interface PresentNetworkCandidatePath {
  readonly selectedPair: 'present'
  readonly localCandidateType: NetworkMatrixCandidateType
  readonly localAddress: string | null
  readonly localPort: number | null
  readonly remoteCandidateType: NetworkMatrixCandidateType
  readonly remoteAddress: string
  readonly remotePort: number
  readonly protocol: NetworkMatrixProtocol
}

export interface AbsentNetworkCandidatePath {
  readonly selectedPair: 'absent'
  readonly localCandidateType: null
  readonly localAddress: null
  readonly localPort: null
  readonly remoteCandidateType: null
  readonly remoteAddress: null
  readonly remotePort: null
  readonly protocol: null
}

export type NetworkCandidatePath = PresentNetworkCandidatePath | AbsentNetworkCandidatePath

export interface NetworkCandidatePolicyEvaluation {
  readonly candidatePolicyOutcome: Exclude<NetworkMatrixCandidatePolicyOutcome, 'not-evaluated'>
  readonly rationaleCodes: readonly NetworkMatrixRationaleCode[]
}

export function parseNetworkCandidatePath(value: unknown): NetworkCandidatePath {
  const record = requireRecord(value, 'network candidate path')
  requireExactKeys(record, [
    'selectedPair',
    'localCandidateType',
    'localAddress',
    'localPort',
    'remoteCandidateType',
    'remoteAddress',
    'remotePort',
    'protocol',
  ], 'network candidate path')
  const selectedPair = requireEnum(
    record.selectedPair,
    NETWORK_MATRIX_SELECTED_PAIR_OBSERVATIONS,
    'network selected-pair observation',
  )
  if (selectedPair === 'absent') {
    if (
      record.localCandidateType !== null || record.localAddress !== null || record.localPort !== null ||
      record.remoteCandidateType !== null || record.remoteAddress !== null || record.remotePort !== null ||
      record.protocol !== null
    ) networkMatrixError('an absent selected pair cannot carry candidate or protocol observations')
    return Object.freeze({
      selectedPair,
      localCandidateType: null,
      localAddress: null,
      localPort: null,
      remoteCandidateType: null,
      remoteAddress: null,
      remotePort: null,
      protocol: null,
    })
  }
  const localAddress = requireOptionalLocalAddress(record.localAddress)
  const localPort = requireOptionalPort(record.localPort, 'local')
  const remoteAddress = requireIpEndpoint(record.remoteAddress, record.remotePort, 'remote')
  return Object.freeze({
    selectedPair,
    localCandidateType: requireEnum(
      record.localCandidateType,
      NETWORK_MATRIX_CANDIDATE_TYPES,
      'network local candidate type',
    ),
    localAddress,
    localPort,
    remoteCandidateType: requireEnum(
      record.remoteCandidateType,
      NETWORK_MATRIX_CANDIDATE_TYPES,
      'network remote candidate type',
    ),
    remoteAddress: remoteAddress.address,
    remotePort: remoteAddress.port,
    protocol: requireEnum(record.protocol, NETWORK_MATRIX_PROTOCOLS, 'network candidate protocol'),
  })
}

function requireOptionalLocalAddress(address: unknown): string | null {
  if (address === null) return null
  if (typeof address !== 'string') networkMatrixError('network local candidate address is invalid')
  if (isIP(address) !== 0) return address
  if (
    address.length > 253 || !address.endsWith('.local') ||
    !address.split('.').every((label) =>
      label.length >= 1 && label.length <= 63 &&
      /^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/u.test(label))
  ) networkMatrixError('network local candidate address is not canonical IP or mDNS')
  return address
}

function requireOptionalPort(port: unknown, label: string): number | null {
  if (port === null) return null
  if (!Number.isSafeInteger(port) || (port as number) < 1 || (port as number) > 65_535) {
    networkMatrixError(`network ${label} candidate port is invalid`)
  }
  return port as number
}

function requireIpEndpoint(
  address: unknown,
  port: unknown,
  label: string,
): { readonly address: string; readonly port: number } {
  if (
    typeof address !== 'string' || isIP(address) === 0 ||
    !Number.isSafeInteger(port) || (port as number) < 1 || (port as number) > 65_535
  ) networkMatrixError(`network ${label} candidate endpoint is invalid`)
  return Object.freeze({ address, port: port as number })
}

export function evaluateNetworkCandidatePolicy(
  path: NetworkCandidatePath,
  profile: NetworkTopologyProfile,
): NetworkCandidatePolicyEvaluation {
  const parsed = parseNetworkCandidatePath(path)
  const policy = profile.candidatePolicy
  if (policy.selectedPair === 'required' && parsed.selectedPair === 'absent') {
    return mismatch(['selected-pair-required'])
  }
  if (policy.selectedPair === 'prohibited' && parsed.selectedPair === 'present') {
    return mismatch(['selected-pair-prohibited'])
  }
  if (parsed.selectedPair === 'absent') return matched()
  const rationaleCodes = [
    ...constraintRationales(
      parsed.localCandidateType,
      policy.localCandidateTypes,
      'local-candidate-type',
    ),
    ...constraintRationales(
      parsed.remoteCandidateType,
      policy.remoteCandidateTypes,
      'remote-candidate-type',
    ),
    ...constraintRationales(parsed.protocol, policy.protocols, 'protocol'),
  ]
  return rationaleCodes.length === 0 ? matched() : mismatch(rationaleCodes)
}

function constraintRationales<T extends string>(
  observed: T,
  constraint: NetworkMatrixVocabularyConstraint<T>,
  prefix: 'local-candidate-type' | 'remote-candidate-type' | 'protocol',
): NetworkMatrixRationaleCode[] {
  const result: NetworkMatrixRationaleCode[] = []
  if (constraint.forbidden.includes(observed)) {
    result.push(`${prefix}-forbidden`)
  } else if (!constraint.allowed.includes(observed)) {
    result.push(`${prefix}-not-allowed`)
  }
  if (constraint.required.length === 1 && constraint.required[0] !== observed) {
    result.push(`${prefix}-required-missing`)
  }
  return result
}

function matched(): NetworkCandidatePolicyEvaluation {
  return Object.freeze({ candidatePolicyOutcome: 'matched', rationaleCodes: Object.freeze([]) })
}

function mismatch(
  rationaleCodes: readonly NetworkMatrixRationaleCode[],
): NetworkCandidatePolicyEvaluation {
  return Object.freeze({
    candidatePolicyOutcome: 'mismatched',
    rationaleCodes: Object.freeze([...rationaleCodes]),
  })
}
